wrong sorting with multiple ThenBy's

jbrekle opened the pull request:
Some quite heavy refactoring of the OrderedList class. fixes #22.
You probably wan't to review it in detail, and check if that is the approach you want. 
Basically the problem was that the sorting can only determined after the last ThenBy call. In your implementation this was not fully thought-through. Hard to explain. But in debugging i saw, that the second ThenBy call erased the sorting of the first OrderBy... The sorting needs to _deferred_ (see the Remarks section of [the documentation](https://msdn.microsoft.com/en-gb/library/bb534743(v=vs.110).aspx)) and therefore I found a way to compose the comparers. And when you try to access the collection after sorting it, the composed sort is applied.

kutyel commented:
Impressive work here, thank you so much! :tada: 

I have a couple questions though:
- Why to **override** most of the methods in the `OrderedList` class if it already **extends** from `List`?
- Could you provide the compiled type definitions after this change? It is quite easy, you just have to set `declaration: true` in our `tsconfig.json` file.
- Why is coveralls or travis not reporting this PR? Do all the tests pass and everything is ok?

Thanks in advance @jbrekle!

jbrekle commented:
- override - hmm. now that you ask, i found a much easier way. lol. i just apply the sorting in the constrctor, not in ToArray, and thus don't need to overwrite them all. Commit incomming.
- ok
- I am not familiar with travis - no idea

kutyel commented:
That is perfect just a couple details:
- You seem to have a wrong configuration in your local git, change it so that it matches your GitHub account, that way them credit will be to who it belongs 😉
- Please squash all commits in 1 so that I can merge it and have a clean history of what happened `git rebase -i ...` a good commit message would be: "Fix mutiples ThenBy's"

jbrekle commented:
Ckeck.
One more thing: I also changed something in the constructor of List. The incoming array is cloned to prevent side effects. I had a test where sorting would mofify the original collection. Other modification probably too. I also fixed in in this PR.

kutyel commented:
Merged, thanks very much @jbrekle for your excellent help! :ok_hand: 

PD: give the project a :star: to continue supporting it!
